### DISCUSSION TITLE
Some way to set environment variables

### DISCUSSION THREAD

**Discussion by hassila** [SwiftPackageIndex/SwiftPackageIndex-Server#2307] [Ideas]
It'd be great to be able to set environment variable in .spi.yml somehow - specifically, I've worked around the requirement of needing `jemalloc` for building the benchmark package here:

https://github.com/ordo-one/package-benchmark/pull/127

So I'd like to set `BENCHMARK_DISABLE_JEMALLOC=1` when building on macOS for SPI (to get the build matrix work for macOS too).

Perhaps there are security implications with that, as a workaround maybe I could ask the above environment to be set for the macOS build machines?

**Comment by daveverwer** [comment]
We’ve thought about this before, specifically about setting some kind of generic `SPI_BUILD` environment variable or something and if I remember correctly we decided not to as it sends us down a route where packages could artificially adjust build results.

I don’t think many people would go to the effort of doing that, but there’s certainly an argument that says we should try and represent the most basic situation for compiling a package, which is what most people who initially try and use it would face, where it would fail without that dependency being installed.

I’m not saying we won’t or shouldn’t do it, but I think it’s worth having that discussion again before we make any changes.

**Reply by hassila** [reply]
Yeah. Just to be clear - in this case setting the variable will remove the dependency and build the package with a subset of functionality disabled.

**Reply by finestructure** [reply]
I think in this case it's worth setting that variable. My understanding is that jemalloc is more like an optimised way of using the package and it's still a useful dependency even without it being enabled.

I'd be in favour of setting `SPI_BUILD` so the package can report a more accurate result.

**Reply by hassila** [reply]
Yes, it you disable `jemalloc` the only thing that happens is that the malloc metrics (which is just a small subset of all metrics) will report 0 as the underlying support for capturing them is missing - 98% of the package is still working and is useful as is (not all users will even use the malloc metrics and might not need them).

**Reply by daveverwer** [reply]
While I agree with this, I do want to play devil's advocate a little.

What if it were flipped and you opted _IN_ to the performant version, rather than out using something specific to the SPI?

That way, you’d also get a package that would work for someone without the dependency installed if they were trying it out. Adding a console log message and a prominent note in the README would also make it hard to miss (especially the console log) that there was a more performant version that could be enabled with a define or an environment variable. 

That way, it wouldn’t need custom tweaking for SPI, and as a bonus side effect you also get a smoother onboarding for people evaluating the package.

**Reply by hassila** [reply]
Fair enough - because I did consider that, and decided against it -  as we would expect the vast majority of users to want to have malloc statistics and wanted to optimise for the common case working out of the box rather than the exception (it is really only in constrained build environments where you can't get jemalloc in where it's a problem).

It'd be a bit of a bummer to have to set up the environment properly for all new users to get valid malloc stat results - 
now they'll get an error that jemalloc is missing (if they didn't read the instructions), do `brew install jemalloc` (or equivalent, once per machine) and then be on the way - so it all boils down to, optimise for the common (99%) case, not the exception. Now, as it happens, SPI builds are one of those 1% constrained environments where we can't install the prerequisite library on the builders easily so here we are...

Now, if we only dynamically could check in Package.swift if the library is available instead of checking an environment variable, we could instead have things 'just work' out of the box...

**Reply by daveverwer** [reply]
Yep, and I won't fight it if we decide a specific SPI build flag is the way to go. I would like to avoid that for as long as possible, but it may be that this is the point where it's needed.

It was just trying to look at it from every angle

**Reply by hassila** [reply]
Well, I can see your rationale for not wanting to add generic environment variable support or specific settings per project as it can be a maintenance burden. 

Feel free to close this if so, I just thought it'd be a way to give more correct package build information in this case (as it is possible to get it working on macOS too trivially, but right now the macOS build looks broken) and thought it was worth pitching.

**Reply by finestructure** [reply]
Hi @hassila , I just wanted to check if what Dave suggested above (inverting the env variable logic such that the default macOS build wouldn't required the env variable to be set) would be an option for you? It seems it would be in the spirit of jemalloc being a performance option.

If that's not an option for you, we can add an env variable (although it's not quite as straightforward as I'd hoped it'd be – it'll require code changes).

**Reply by hassila** [reply]
Hi, thanks - sorry for late reply have been out of office - just saw the related case with the SPI environment variable - thanks, much appreciated (wouldn't make sense to change the default IMHO as it's not really a performance related option, but rather fundamental functionality that I'd expect the vast majority of users to want to have....). Looking forward to try out the env var, thank you!

**Comment by heckj** [comment]
Yep, I figured you'd fenced it off and disabled the code with an environment variable trigger.

**Reply by hassila** [reply]
Actually I fenced off the package dependency and the target dependency - the code can then simply use canImport() appropriately.

**Comment by finestructure** [comment]
This is implemented as `SPI_BUILD` (see #2336)
